Fix the tests

asmeurer opened the pull request:
Fixes #7.

asmeurer commented:
I have fixed all the test failures, except for a bunch of timeout failures from pexpect. Here is an example failure

```

______________________ test_ipython_tab_fail_1[readline] _______________________
frontend = 'readline'
    def test_ipython_tab_fail_1(frontend):
        # Test that our test harness works for tab when it should match nothing.
        ipython("""
            In [1]: import os
            In [2]: os.foo27817796\t()
            ---------------------------------------------------------------------------
            AttributeError                            Traceback (most recent call last)
            <ipython-input> in <module>()
            AttributeError: 'module' object has no attribute 'foo27817796'
>       """, frontend=frontend)
tests/test_interactive.py:1400: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
self = <contextlib.GeneratorContextManager object at 0x7f3264239410>
type = <class 'pexpect.exceptions.TIMEOUT'>
value = TIMEOUT("Timeout exceeded.\n<tests.test_interactive.MySpawn object at 0x7f3264...8\\\\x1b\\\\[K\\\\x08\\\\x1b\\\\[K\\\\x08\\\\x1b\\\\[K\\\\x08\\\\x1b\\\\[K')",)
traceback = <traceback object at 0x7f3262f07440>
    def __exit__(self, type, value, traceback):
        if type is None:
            try:
                self.gen.next()
            except StopIteration:
                return
            else:
                raise RuntimeError("generator didn't stop")
        else:
            if value is None:
                # Need to force instantiation so we can reliably
                # tell if we get the same exception back
                value = type()
            try:
>               self.gen.throw(type, value, traceback)
E               ExpectError: TIMEOUT in /usr/bin/env IPYTHONDIR=/tmp/pyflyby_test_ipython_itY8nO.tmp INPUTRC= ipython --no-confirm-exit --no-banner --InteractiveShell.autoindent=False --TerminalIPythonApp.interactive_shell_class=rlipython.TerminalInteractiveShell
/opt/python/2.7.14/lib/python2.7/contextlib.py:35: ExpectError
----------------------------- Captured stdout call -----------------------------
Input:
    import os
    os.foo27817796	()
# Spawning: env IPYTHONDIR=/tmp/pyflyby_test_ipython_itY8nO.tmp INPUTRC= ipython --no-confirm-exit --no-banner --InteractiveShell.autoindent=False --TerminalIPythonApp.interactive_shell_class=rlipython.TerminalInteractiveShell
Error: TIMEOUT
Output so far:
    In [1]: import os
    In [2]: os.foo27817796
Error details:
    Timeout exceeded.
    <tests.test_interactive.MySpawn object at 0x7f3264239e50>
    command: /usr/bin/env
    args: ['/usr/bin/env', 'IPYTHONDIR=/tmp/pyflyby_test_ipython_itY8nO.tmp', 'INPUTRC=', 'ipython', '--no-confirm-exit', '--no-banner', '--InteractiveShell.autoindent=False', '--TerminalIPythonApp.interactive_shell_class=rlipython.TerminalInteractiveShell']
    buffer (last 100 chars): '\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08'
    before (last 100 chars): '\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08'
    after: <class 'pexpect.exceptions.TIMEOUT'>
    match: None
    match_index: None
    exitstatus: None
    flag_eof: False
    pid: 3978
    child_fd: 16
    closed: False
    timeout: 10.0
    delimiter: <class 'pexpect.exceptions.EOF'>
    logfile: None
    logfile_read: None
    logfile_send: None
    maxread: 2000
    ignorecase: False
    searchwindowsize: None
    delaybeforesend: 0.0
    delayafterclose: 0.1
    delayafterterminate: 0.1
    searcher: searcher_re:
        0: re.compile('\\x1b\\[16D')
        1: re.compile('\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K')
```

Various tests seem to fail this way and I'm not sure what causes the failure. The only relationship I can see is that they all use the readline frontend, but not every readline frontend test fails. Raising the timeout time does not fix the issue.

asmeurer commented:
Maybe the common thread is that they all have a tab in the input.

asmeurer commented:
There's also random failures with prompt toolkit where inputs aren't shown sometimes (https://travis-ci.com/asmeurer/pyflyby/jobs/188034963#L872, https://travis-ci.com/asmeurer/pyflyby/jobs/188281681#L1127). 

It's possible these would go away with Python 3, since prompt toolkit 2.0 and the newest IPython both have a lot of changes.

quarl commented:
Hm, it works for me.
I tried two different environments and it worked in both.


machine1$ py.test -k test_ipython_tab_fail_1 -v
platform linux2 -- Python 2.7.16rc1, pytest-3.10.1, py-1.7.0, pluggy-0.8.0 -- /usr/bin/python2
IPython 5.8.0
pyflyby 1.2.5 from /home/quarl/proj/pyflyby/lib/python/pyflyby
cachedir: .pytest_cache
rootdir: /home/quarl/proj/pyflyby, inifile:
plugins: xdist-1.26.1, forked-1.0.1
collected 1360 items / 1358 deselected

tests/test_interactive.py::test_ipython_tab_fail_1[readline] PASSED                                                                                                                                                                                                     [ 50%]
tests/test_interactive.py::test_ipython_tab_fail_1[prompt_toolkit] PASSED                                                                                                                                                                                               [100%]



machine2$ py.test -k test_ipython_tab_fail_1 -v
platform linux2 -- Python 2.7.15, pytest-3.10.1, py-1.7.0, pluggy-0.8.0 -- /usr/local/python/python-2.7/std/bin/python2.7
IPython 5.8.0
pyflyby 1.2.5 from /u/chenk/proj/pyflyby2/lib/python/pyflyby
cachedir: .pytest_cache
metadata: {'Python': '2.7.15', 'Platform': 'Linux-3.10.0-957.5.1.el7.7.deshaw.x86_64-x86_64-with-redhat-7.6-Maipo', 'Packages': {'py': '1.7.0', 'pytest': '3.10.1', 'pluggy': '0.8.0'}, 'Plugins': {'cov': '2.5.1', 'nbval': '0.9.0', 'benchmark': '3.1.1', 'xdist': '1.22.2',
'json': '0.4.0', 'html': '1.15.1', 'hypothesis': '3.57.0', 'ipynb': '1.1.0', 'forked': '0.2', 'metadata': '1.7.0'}}
benchmark: 3.1.1 (defaults: timer=time.time disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /u/chenk/proj/pyflyby2, inifile:
plugins: xdist-1.22.2, metadata-1.7.0, json-0.4.0, ipynb-1.1.0, html-1.15.1, forked-0.2, cov-2.5.1, benchmark-3.1.1, nbval-0.9.0, hypothesis-3.57.0
collected 1360 items / 1358 deselected

tests/test_interactive.py::test_ipython_tab_fail_1[readline] PASSED                                                                                                                                                                                                     [ 50%]
tests/test_interactive.py::test_ipython_tab_fail_1[prompt_toolkit] PASSED                                                                                                                                                                                               [100%]



From: Aaron Meurer <notifications@github.com>
Sent: Wednesday, March 27, 2019 14:46
To: pyflyby/pyflyby <pyflyby@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [pyflyby/pyflyby] Fix the tests (#8)


I have fixed all the test failures, except for a bunch of timeout failures from pexpect. Here is an example failure



______________________ test_ipython_tab_fail_1[readline] _______________________

frontend = 'readline'

    def test_ipython_tab_fail_1(frontend):

        # Test that our test harness works for tab when it should match nothing.

        ipython("""

            In [1]: import os

            In [2]: os.foo27817796\t()

            ---------------------------------------------------------------------------

            AttributeError                            Traceback (most recent call last)

            <ipython-input> in <module>()

            AttributeError: 'module' object has no attribute 'foo27817796'

>       """, frontend=frontend)

tests/test_interactive.py:1400:

_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <contextlib.GeneratorContextManager object at 0x7f3264239410>

type = <class 'pexpect.exceptions.TIMEOUT'>

value = TIMEOUT("Timeout exceeded.\n<tests.test_interactive.MySpawn object at 0x7f3264...8\\\\x1b\\\\[K\\\\x08\\\\x1b\\\\[K\\\\x08\\\\x1b\\\\[K\\\\x08\\\\x1b\\\\[K')",)

traceback = <traceback object at 0x7f3262f07440>

    def __exit__(self, type, value, traceback):

        if type is None:

            try:

                self.gen.next()

            except StopIteration:

                return

            else:

                raise RuntimeError("generator didn't stop")

        else:

            if value is None:

                # Need to force instantiation so we can reliably

                # tell if we get the same exception back

                value = type()

            try:

>               self.gen.throw(type, value, traceback)

E               ExpectError: TIMEOUT in /usr/bin/env IPYTHONDIR=/tmp/pyflyby_test_ipython_itY8nO.tmp INPUTRC= ipython --no-confirm-exit --no-banner --InteractiveShell.autoindent=False --TerminalIPythonApp.interactive_shell_class=rlipython.TerminalInteractiveShell

/opt/python/2.7.14/lib/python2.7/contextlib.py:35: ExpectError

----------------------------- Captured stdout call -----------------------------

Input:

    import os

    os.foo27817796     ()

# Spawning: env IPYTHONDIR=/tmp/pyflyby_test_ipython_itY8nO.tmp INPUTRC= ipython --no-confirm-exit --no-banner --InteractiveShell.autoindent=False --TerminalIPythonApp.interactive_shell_class=rlipython.TerminalInteractiveShell

Error: TIMEOUT

Output so far:

    In [1]: import os

    In [2]: os.foo27817796

Error details:

    Timeout exceeded.

    <tests.test_interactive.MySpawn object at 0x7f3264239e50>

    command: /usr/bin/env

    args: ['/usr/bin/env', 'IPYTHONDIR=/tmp/pyflyby_test_ipython_itY8nO.tmp', 'INPUTRC=', 'ipython', '--no-confirm-exit', '--no-banner', '--InteractiveShell.autoindent=False', '--TerminalIPythonApp.interactive_shell_class=rlipython.TerminalInteractiveShell']

    buffer (last 100 chars): '\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08'

    before (last 100 chars): '\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08\x08'

    after: <class 'pexpect.exceptions.TIMEOUT'>

    match: None

    match_index: None

    exitstatus: None

    flag_eof: False

    pid: 3978

    child_fd: 16

    closed: False

    timeout: 10.0

    delimiter: <class 'pexpect.exceptions.EOF'>

    logfile: None

    logfile_read: None

    logfile_send: None

    maxread: 2000

    ignorecase: False

    searchwindowsize: None

    delaybeforesend: 0.0

    delayafterclose: 0.1

    delayafterterminate: 0.1

    searcher: searcher_re:

        0: re.compile('\\x1b\\[16D')

        1: re.compile('\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K\\x08\\x1b\\[K')

Various tests seem to fail this way and I'm not sure what causes the failure. The only relationship I can see is that they all use the readline frontend, but not every readline frontend test fails. Raising the timeout time does not fix the issue.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/pyflyby/pyflyby/pull/8#issuecomment-477299831>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AADDZ_8G_AqojM-kNdQHFfEJF6EGOgAVks5va7xkgaJpZM4cARkj>.

quarl commented:
Re the intermittent failures in CI:
We might need to adjust the sleep(0.02) or the timeout=0.05 upwards.  Maybe the CI machine gets more overloaded and occasionally pauses for 0.05s.
(Or better yet, find an approach to testing prompt_toolkit that doesn’t rely on sleeping.)


From: Aaron Meurer <notifications@github.com>
Sent: Wednesday, March 27, 2019 17:01
To: pyflyby/pyflyby <pyflyby@noreply.github.com>
Cc: Subscribed <subscribed@noreply.github.com>
Subject: Re: [pyflyby/pyflyby] Fix the tests (#8)


There's also random failures with prompt toolkit where inputs aren't shown sometimes (https://travis-ci.com/asmeurer/pyflyby/jobs/188034963#L872, https://travis-ci.com/asmeurer/pyflyby/jobs/188281681#L1127).

It's possible these would go away with Python 3, since prompt toolkit 2.0 and the newest IPython both have a lot of changes.

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub<https://github.com/pyflyby/pyflyby/pull/8#issuecomment-477345819>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AADDZ1aEGCXkI24JvJNnKG8u8Yhzyl0xks5va9wWgaJpZM4cARkj>.

asmeurer commented:
sleep(0.02) only relates to autoindent as far as I can tell, which isn't the case for the ones I've seen. So I guess it is the 0.05 timeout. 

I would ideally like to fix the other error first, as it's kind of hard to find the non-timeout errors in the test output. I've had luck on Travis doing a retries with sleep times like `0.1, 1.05, 2` (linear increase) so that the tests still run fast locally where things aren't slow.

quarl commented:
Escalating timeouts sounds good.  We might even be able to shorten the initial timeout, then.

asmeurer commented:
I fixed some of the timeouts (they were my fault from removing the clear line ANSI escape sequence, which is used in the tab completion checking code).

The other timeouts, such as `test_py_console_1`, also occur for me in `master`, so for those at least I don't think I caused them :)

asmeurer commented:
Oh I see, it's trying to use the `py` utility from `bin/`, but it isn't finding it.

asmeurer commented:
Looks like I accidentally committed some code to master and pushed to master. Sorry about that. I have set my remote URL to git:// so it won't happen again. We can also enable branch protection here which would make it impossible to push to master (all code has to go through pull requests).

asmeurer commented:
Incidentally, the [tests pass now](https://travis-ci.com/asmeurer/pyflyby/builds/106235834). I think there might still be random failures from the timeouts as described above. 

Travis isn't building this PR. I guess it doesn't trigger on PRs that were opened before it was set up.

asmeurer commented:
I guess it is running now. I don't know it wasn't before.

asmeurer commented:
I set the `_wait_for_output` timeout up to 60 seconds and there are still failures on Travis. Hopefully the failures are really being caused by something else and it's not the case that the timeout needs to be that large.

quarl commented:
Some of the tests currently have the unfortunate property that if it doesn't see the expected output, then it waits forever until timeout; and unfortunately currently the test output doesn't clearly indicate what it was waiting for and what it received so far.

asmeurer commented:
Yes, I noticed that. I think the debug output helps here. I thought it was indicating that _wait_for_output was the culprit, but I guess it must not be. Or perhaps it is both the timeout and the sleep time that matter.

asmeurer commented:
I've added automatic retries to the randomly failing tests. I can't figure out exactly what timeout causes them to fail, so I didn't adjust that. There are about a dozen or so tests that fail randomly and each run has about 1 or 2 fail each time, so a simple retry should be enough. I've only added retrying to tests that I've observed fail from the "lost input" issue on Travis. So there may still be some randomly failing tests that need the @retry decorator.

The @retry decorator that was there didn't work. pytest fixtures do not work so well with decorators. I couldn't figure out how to make it work, so I replaced it with the flaky pytest plugin, which is what the [pytest docs](https://docs.pytest.org/en/latest/flaky.html) recommend using. 

I will restart the Travis run on my fork a few times to make sure that there aren't any missing ones. But otherwise, I think this is ready to be reviewed and merged.

asmeurer commented:
This is ready.
